internal/runner/reattach.go

```diff
@@ -0,0 +1,262 @@
+		// Should never happen.
+		panic(err)
+	}
+
+	return resp
+}
+
+func (c *reattachClient) Context() context.Context {
+	var resp context.Context
+	err := c.do(func(client pb.Waypoint_RunnerJobStreamClient) error {
+		resp = client.Context()
+		return nil
+	})
+	if err != nil {
+		// Should never happen.
+		panic(err)
+	}
+
+	return resp
+}
+
+func (c *reattachClient) CloseSend() error {
+	return c.do(func(client pb.Waypoint_RunnerJobStreamClient) error {
+		return client.CloseSend()
+	})
+}
+
+func (c *reattachClient) SendMsg(m interface{}) error {
+	return c.do(func(client pb.Waypoint_RunnerJobStreamClient) error {
+		return client.SendMsg(m)
+	})
+}
+
+func (c *reattachClient) RecvMsg(m interface{}) error {
+	return c.do(func(client pb.Waypoint_RunnerJobStreamClient) error {
+		return client.RecvMsg(m)
+	})
+}
+
+func (c *reattachClient) do(f func(client pb.Waypoint_RunnerJobStreamClient) error) error {
+	// Shorthand for common attribute access
+	log := c.log
+	r := c.runner
+
+	// Get our configuration state value. We use this so that we can detect
+	// when we've reconnected during failures.
+	stateGen := r.readState(&r.stateConfig)
+
+	// We use this as a way to avoid having unlocks everywhere since we're
+	// using a loop.
+	locked := false
+	defer func() {
+		if locked {
+			c.clientLock.Unlock()
+			locked = false
+		}
+	}()
+
+	for {
+		// If we don't have a lock (our first time through), then we grab it
+		// so we can get our client. If we have a lock (a connection retry)
+		// then we avoid this. We always unlock after this.
+		if !locked {
+			c.clientLock.Lock()
+		}
+
+		// Get our current client
+		client := c.client
+		clientGen := c.clientGen
+		c.clientLock.Unlock()
+		locked = false
+
+		// Do operation
+		err := f(client)
+		if err == nil {
+			// No error means we're done.
+			return nil
+		}
+
+		// We attempt a reconnect on Unavailable and NotFound but otherwise
+		// we return the error as-is.
+		if status.Code(err) != codes.Unavailable &&
+			status.Code(err) != codes.NotFound {
+			return err
+		}
+		log.Warn("server down during job API, will reconnect")
+
+		// Lock so we only ever reconnect once at a time.
+		c.clientLock.Lock()
+		locked = true
+		if c.clientGen != clientGen {
+			// If the client gen changed, we already reconnected.
+			continue
+		}
+
+		// Since this is a disconnect, we have to wait for our
+		// RunnerConfig stream to re-establish. We wait for the config
+		// generation to increment.
+		log.Info("waiting for runner to re-register before reconnecting")
+		if r.waitStateGreater(&r.stateConfig, stateGen) {
+			return status.Error(codes.Internal, "early exit while waiting for reconnect")
+		}
+
+		// Get the new state generation in case we get another disconnect
+		// we can continue retrying and detecting new connections.
+		stateGen = r.readState(&r.stateConfig)
+
+		log.Info("opening job stream")
+		client, err = r.client.RunnerJobStream(context.Background(), grpc.WaitForReady(true))
```

Comment:
How long with this block for if the server is down? Indefinitely? Since we're using a new context, can you add a comment about this lines "blocking mode"?

Reply:
This was a mistake, now fixed to use the context given to `accept` so the lifecycle matches that.